### ISSUE TITLE
How hard to double resolution?

### ISSUE THREAD

**Issue by frosty3907** [deepfakes/faceswap-playground#122]
The main issue in my eyes with the current version is the low resolution of the output faces, how difficult is it to simply allow a command line switch to double the resolution? Yes I realize this would slow things down but in my mind it would be well worth it.

**Comment by Kirin-kun** [comment]
I don't think you understand how it works. You won't have more crisp faces with higher resolution because the neural network averages all the faces. You won't have more details that way.

Maybe with https://en.wikipedia.org/wiki/Face_hallucination

**Comment by bryanlyon** [comment]
It's also not possible to simply double the resolution, it requires a model that is 2^2 (4x) the size in vram to get similar results.  This means that the Original model would require at least 16gb of ram with a proper 128 resolution.  There are tricks to make this work with less ram (which is what GAN128 is using).

My advice, if you think you NEED higher resolution is to try out the GAN128.

I think the eyes will still be problematic since the GAN does not do special processing for eyes and treats them like any other part of the face, learning the average.  Since eyes are so important for humans to recognize faces, we might need to make a custom eye processing.  This is something I've been working on for the last couple weeks (I have it in a separate codebase while I experiment.  Once I find a good system for eyes, I'll add a plugin for faceswap.)

**Comment by frosty3907** [comment]
My 1080ti has 11GB VRAM, system RAM is 16GB and although you're right, I don't understand how the system works- it seems if we can get 64x64 w/ a 2GB card an 11GB card should be able to handle 128x128.

It would stop the teeth being all white as well as allowing shots other than extreme long shots too which would be great.

Anyway since I can't contribute I'll leave it at that, but it is my opinion we need to get better final resolution, somehow or other (could you use low res proxies and then sub in higher resolution images during the warping/converting stage?).

**Comment by bryanlyon** [comment]
Lowmem is the only model that can run on a 2gb card.  It could be possible to increase it to 128, but that's something for future research.  However, if you're willing to try a new model,, why not try GAN128?  It's a 128x128 model that we're currently testing and developing.

Even with much higher resolution though, teeth will never be clear.  This is because of how the AI works.  It's not looking at delineating features (such as the space between teeth) it's looking to fool itself that two photos are of the same person, one blurred and really that person and one that is converted from another person's blurred photo.  Since it's getting blurred photos as input, it doesn't even see teeth gaps or much eye detail.  This is why I'm working on an external eye replacement procedure to update the eyes better.  Teeth are similar and wont be clean, even with extreme resolutions.

**Comment by iperov** [comment]
> The problem is the loss function.

Autoencoder principle will ALWAYS average samples regardless on loss fuction.
So let feed it only useful samples, because why feed it with -90 angles of src, if this angle not exist in dst ? ANd I implemented it :p

`How hard to double resolution?`
I done it.

my improved original 64:
![38348898-97b3e920-38b5-11e8-8c05-7421aed78ef6 1](https://user-images.githubusercontent.com/8076202/38349232-21a5ea6a-38b7-11e8-8f01-4d0d19303d02.png)

full face 128 with regular 'mae' loss
(and still training more):
![38348901-9fc975da-38b5-11e8-9300-7bda415e03cd 1](https://user-images.githubusercontent.com/8076202/38349253-385299c0-38b7-11e8-826d-ffb8eb651542.png)

**Comment by Kirin-kun** [comment]
@iperov Do you generate the angles that don't exist in dst?

I understand the idea of not using a face angle that won't be used anyway, but what about angles you don't have?

Something like what's mentioned in [this paper](https://www.researchgate.net/publication/278790083_Cascaded_Collaborative_Regression_for_Robust_Facial_Landmark_Detection_Trained_Using_a_Mixture_of_Synthetic_and_Real_Images_With_Dynamic_Weighting) might be useful, but it's not trivial.

That would make training data more useful and it would need only a few pictures of src face.

**Comment by iperov** [comment]
1) NN predicts angles that you dont have

2) I have ALL angles. My Putin src faceset 18000 photos.

**Comment by iperov** [comment]
I dont need to calculate yaw angle exactly. Point difference is enough.

**Comment by iperov** [comment]
my algorithm:
```
            -yaw_value ... +yaw_value [0..255]
src (Cage): [|||||||||||||||||||||||||]
dst (Trum): [    |   | |              ]
       feed:    ^^^ ^^^^^
```

**Comment by iperov** [comment]
my perfect improved DFaker model

![38381072-ba81e270-3916-11e8-8ea3-71bf9131bfcf 1](https://user-images.githubusercontent.com/8076202/38386832-1110051c-3927-11e8-8338-0c15d5b07123.gif)


I think I made BEST face swapper in the internet.

**Comment by frosty3907** [comment]
How does it look in motion?

**Comment by frosty3907** [comment]
interesting, is it available for testing?

**Comment by iperov** [comment]
not yet

**Comment by iperov** [comment]
> what about angles you don't have?

Implemented search nearest direct and mirror yaw.
If mirror yaw found, landmarks and sample flipped.
In this case 'flip' chance in model's training generator removed.
This makes not feeding NN by useless flipped samples which will be not used in convert.
And as we know: less samples - sharper result.
